File: 백준/Silver/15988. 1， 2， 3 더하기 3/1， 2， 3 더하기 3.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const T = Number(input[0])
const MAX = 1_000_000
const DIVIDE = 1_000_000_009
const dp = Array(MAX + 1).fill(0)
dp[1] = 1
dp[2] = 2
dp[3] = 4

for (let i = 4; i <= MAX; i++) {
  dp[i] = (dp[i - 1] + dp[i - 2] + dp[i - 3]) % DIVIDE
}

for (let tc = 1; tc < T + 1; tc++) {
  const target = Number(input[tc])
  console.log(dp[target])
}

File: 백준/Silver/15988. 1， 2， 3 더하기 3/1， 2， 3 더하기 3.py
import sys
input = sys.stdin.readline

dp = [1,2,4,7]
for tc in range(int(input())):
    n = int(input())
    for i in range(len(dp), n):
        dp.append((dp[-3] + dp[-2] + dp[-1]) % 1000000009)
    print(dp[n-1])
